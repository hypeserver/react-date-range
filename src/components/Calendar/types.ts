import React from "react";


export interface Scroll {
  enabled: boolean
  monthWidth?: number | string
  monthHeight?: number | string
  longMonthHeight?: number | string
  // FIXME
  calendarWidth?: any
  calendarHeight?: any
}

export enum CalendarDirection {
  VERTICAL = 'vertical',
  HORIZONTAL = 'horizontal'
}


export enum ChangeDateMode {
  SET = 'set',
  MONTH_OFFSET = 'monthOffset',
  SET_MONTH = 'setMonth',
  SET_YEAR = 'setYear'
}


export interface AriaLabelsShape {
  dateInput: { startDate: string, endDate: string }[]
  monthPicker: string,
  yearPicker: string,
  prevButton: string,
  nextButton: string,
}
