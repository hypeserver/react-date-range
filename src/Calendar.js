import React, { Component, PropTypes } from 'react';
import moment from 'moment';
import parseInput from './utils/parseInput.js';
import DayCell from './DayCell.js';
import LangDic from './LangDic.js';
import getTheme, { defaultClasses } from './styles.js';

function checkRange(dayMoment, range) {
  return (
    dayMoment.isBetween(range['startDate'], range['endDate']) ||
    dayMoment.isBetween(range['endDate'], range['startDate'])
  )
}

function checkStartEdge(dayMoment, range) {
  const { startDate } = range;

  return dayMoment.startOf('day').isSame(startDate.startOf('day'));
}

function checkEndEdge(dayMoment, range) {
  const { endDate } = range;

  return dayMoment.endOf('day').isSame(endDate.endOf('day'));
}

function isOusideMinMax(dayMoment, minDate, maxDate, format) {
  return (
    (minDate && dayMoment.isBefore(parseInput(minDate, format, 'startOf'))) ||
    (maxDate && dayMoment.isAfter(parseInput(maxDate, format, 'endOf')))
  )
}

class Calendar extends Component {

  constructor(props, context) {
    super(props, context);

    const { format, range, theme, offset, firstDayOfWeek, locale, shownDate } = props;

    if(locale) {
      moment.locale(locale);
    }

    const date = parseInput(props.date, format, 'startOf')
    const state = {
      date,
      shownDate : (shownDate || range && range['endDate'] || date).clone().add(offset, 'months'),
      firstDayOfWeek: (firstDayOfWeek || moment.localeData().firstDayOfWeek()),
    }

    this.state  = state;
    this.styles = getTheme(theme);
  }

  componentDidMount() {
    const { onInit } = this.props;
    onInit && onInit(this.state.date);
  }

  componentWillReceiveProps(nextProps) {
    const { range, offset } = nextProps;
    const oldRange = this.props.oldRange;

    if ((range && range['endDate'] && !range['endDate'].isSame(range['startDate'], "day")) || (oldRange && !oldRange["startDate"].isSame(range["startDate"]))) {
      this.setState({ shownDate : range['endDate'].clone().add(offset, 'months') })
    }
  }

  getShownDate() {
    const { link, offset } = this.props;

    const shownDate = (link) ? link.clone().add(offset, 'months') : this.state.shownDate;

    return shownDate;
  }

  handleSelect(newDate) {
    const { link, onChange } = this.props;
    const { date } = this.state;

    onChange && onChange(newDate, Calendar);

    if (!link) {
      this.setState({ date : newDate });
    }
  }

  changeMonth(direction, event) {
    event.preventDefault();
    const { link, linkCB } = this.props;

    if (link && linkCB) {
      return linkCB(direction);
    }

    const current  = this.state.shownDate.month();
    const newMonth = this.state.shownDate.clone().add(direction, 'months');

    this.setState({
      shownDate : newMonth
    });
  }

  renderMonthAndYear(classes) {
    const shownDate       = this.getShownDate();
    let month           = moment.months(shownDate.month());
    const year            = shownDate.year();
    const { styles }      = this;
    const { onlyClasses, lang, showMonthArrow} = this.props;

    let monthLower = month.toLowerCase()
    month = (lang && LangDic[lang] && LangDic[lang][monthLower]) ? LangDic[lang][monthLower] : month;

    return (
      <div style={onlyClasses ? undefined : styles['MonthAndYear']} className={classes.monthAndYearWrapper}>
        {
          showMonthArrow ?
          <button
            type="button"
            style={onlyClasses ? undefined : { ...styles['MonthButton'], float : 'left' }}
            className={classes.prevButton}
            onClick={this.changeMonth.bind(this, -1)}>
            <i style={onlyClasses ? undefined : { ...styles['MonthArrow'], ...styles['MonthArrowPrev'] }}></i>
          </button> : null
        }
        <span>
          <span className={classes.month}>{month}</span>
          <span className={classes.monthAndYearDivider}> - </span>
          <span className={classes.year}>{year}</span>
        </span>
        {
          showMonthArrow ?
          <button
            type="button"
            style={onlyClasses ? undefined : { ...styles['MonthButton'], float : 'right' }}
            className={classes.nextButton}
            onClick={this.changeMonth.bind(this, +1)}>
            <i style={onlyClasses ? undefined : { ...styles['MonthArrow'], ...styles['MonthArrowNext'] }}></i>
          </button> : null
        }
      </div>
    )
  }

  renderWeekdays(classes) {
    const dow             = this.state.firstDayOfWeek;
    const weekdays        = [];
    const { styles }      = this;
    const { onlyClasses, lang } = this.props;

    for (let i = dow; i < 7 + dow; i++) {
      let day = moment.weekdaysMin(i);
      let dayLower = day.toLowerCase();
      day = (lang && LangDic[lang] && LangDic[lang][dayLower]) ? LangDic[lang][dayLower] : day;
      weekdays.push(
        <span style={onlyClasses ? undefined : styles['Weekday']} className={classes.weekDay} key={i + day}>{day}</span>
      );
    }

    return weekdays;
  }

  renderDays(classes) {
    // TODO: Split this logic into smaller chunks
    const { styles }               = this;

    const { range, maxRange, minDate, maxDate, format, onlyClasses, disableDaysBeforeToday, specialDays } = this.props;

    const shownDate                = this.getShownDate();
    const { date, firstDayOfWeek } = this.state;
    const dateUnix                 = date.unix();

    const monthNumber              = shownDate.month();
    const dayCount                 = shownDate.daysInMonth();
    const startOfMonth             = shownDate.clone().startOf('month').isoWeekday();

    const lastMonth                = shownDate.clone().month(monthNumber - 1);
    const lastMonthNumber          = lastMonth.month();
    const lastMonthDayCount        = lastMonth.daysInMonth();

    const nextMonth                = shownDate.clone().month(monthNumber + 1);
    const nextMonthNumber          = nextMonth.month();

    const days                     = [];

    // Previous month's days
    const diff = (Math.abs(firstDayOfWeek - (startOfMonth + 7)) % 7);
    for (let i = diff-1; i >= 0; i--) {
      const dayMoment  = lastMonth.clone().date(lastMonthDayCount - i);
      days.push({ dayMoment, isPassive : true });
    }

    // Current month's days
    for (let i = 1; i <= dayCount; i++) {
      const dayMoment  = shownDate.clone().date(i);
      // set days before today to isPassive
      var _today = moment()
      if (disableDaysBeforeToday && Number(dayMoment.diff(_today,"days")) <= -1) {
        days.push({ dayMoment ,isPassive:true});
      } else {
        days.push({ dayMoment });
      }
    }

    // Next month's days
    const remainingCells = 42 - days.length; // 42cells = 7days * 6rows
    for (let i = 1; i <= remainingCells; i++ ) {
      const dayMoment  = nextMonth.clone().date(i);
      days.push({ dayMoment, isPassive : true });
    }

    const today = moment().startOf('day');
    return days.map((data, index) => {
      const { dayMoment, isPassive } = data;
      const isSelected    = !range && (dayMoment.unix() === dateUnix);
      const isInRange     = range && checkRange(dayMoment, range);
      const isStartEdge   = range && checkStartEdge(dayMoment, range);
      const isEndEdge     = range && checkEndEdge(dayMoment, range);
      const isEdge        = isStartEdge || isEndEdge;
      const isToday       = today.isSame(dayMoment);
      const isSunday      = dayMoment.day() === 0;
      const isSpecialDay  = specialDays && specialDays.some((specialDay) => {
        return dayMoment.endOf('day').isSame(specialDay.date.endOf('day'));
      });
      const isOutsideMinMax = isOusideMinMax(dayMoment, minDate, maxDate, format);
      const isDisabled = maxRange && dayMoment.startOf('day').isAfter(range.startDate.clone().add(maxRange));

      return (
        <DayCell
          onSelect={ this.handleSelect.bind(this) }
          { ...data }
          theme={ styles }
          isStartEdge = { isStartEdge }
          isEndEdge = { isEndEdge }
          isSelected={ isSelected || isEdge }
          isInRange={ isInRange }
          isSunday={ isSunday }
          isSpecialDay={ isSpecialDay }
          isToday={ isToday }
          key={ index }
          isPassive = { isPassive || isOutsideMinMax }
          isDisabled = { isDisabled }
          onlyClasses = { onlyClasses }
          classNames = { classes }
        />
      );
    })
  }

  render() {
    const { styles } = this;
    const { onlyClasses, classNames } = this.props;

    const classes = { ...defaultClasses, ...classNames };

    return (
      <div style={onlyClasses ? undefined : { ...styles['Calendar'], ...this.props.style }} className={classes.calendar}>
        <div className={classes.monthAndYear}>{ this.renderMonthAndYear(classes) }</div>
        <div className={classes.weekDays}>{ this.renderWeekdays(classes) }</div>
        <div className={classes.days}>{ this.renderDays(classes) }</div>
      </div>
    )
  }
}

Calendar.defaultProps = {
  format      : 'DD/MM/YYYY',
  theme       : {},
  showMonthArrow: true,
  disableDaysBeforeToday: false,
  onlyClasses : false,
  classNames  : {},
  specialDays : [],
}

Calendar.propTypes = {
  showMonthArrow : PropTypes.bool,
  disableDaysBeforeToday : PropTypes.bool,
  lang           : PropTypes.string,
  sets           : PropTypes.string,
  range          : PropTypes.shape({
    startDate    : PropTypes.object,
    endDate      : PropTypes.object
  }),
  minDate        : PropTypes.oneOfType([PropTypes.object, PropTypes.func, PropTypes.string]),
  maxDate        : PropTypes.oneOfType([PropTypes.object, PropTypes.func, PropTypes.string]),
  date           : PropTypes.oneOfType([PropTypes.object, PropTypes.string, PropTypes.func]),
  format         : PropTypes.string.isRequired,
  firstDayOfWeek : PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  onChange       : PropTypes.func,
  onInit         : PropTypes.func,
  link           : PropTypes.oneOfType([PropTypes.shape({
    startDate    : PropTypes.object,
    endDate      : PropTypes.object,
  }), PropTypes.bool]),
  linkCB         : PropTypes.func,
  theme          : PropTypes.object,
  onlyClasses    : PropTypes.bool,
  specialDays    : PropTypes.array,
  classNames     : PropTypes.object,
  locale         : PropTypes.string
}

export default Calendar;
