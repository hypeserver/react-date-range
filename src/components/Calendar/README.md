#### Example: Internationalization

```jsx inside Markdown
import * as locales from '@umakantp/react-date-range/dist/locale';
import { useState } from 'react';

const nameMapper = {
  ar: 'Arabic',
  bg: 'Bulgarian',
  ca: 'Catalan',
  cs: 'Czech',
  cy: 'Welsh',
  da: 'Danish',
  de: 'German',
  el: 'Greek',
  enGB: 'English (United Kingdom)',
  enUS: 'English (United States)',
  eo: 'Esperanto',
  es: 'Spanish',
  et: 'Estonian',
  faIR: 'Persian',
  fi: 'Finnish',
  fil: 'Filipino',
  fr: 'French',
  hi: 'Hindi',
  hr: 'Croatian',
  hu: 'Hungarian',
  hy: 'Armenian',
  id: 'Indonesian',
  is: 'Icelandic',
  it: 'Italian',
  ja: 'Japanese',
  ka: 'Georgian',
  ko: 'Korean',
  lt: 'Lithuanian',
  lv: 'Latvian',
  mk: 'Macedonian',
  nb: 'Norwegian BokmÃ¥l',
  nl: 'Dutch',
  pl: 'Polish',
  pt: 'Portuguese',
  ro: 'Romanian',
  ru: 'Russian',
  sk: 'Slovak',
  sl: 'Slovenian',
  sr: 'Serbian',
  sv: 'Swedish',
  th: 'Thai',
  tr: 'Turkish',
  uk: 'Ukrainian',
  vi: 'Vietnamese',
  zhCN: 'Chinese Simplified',
  zhTW: 'Chinese Traditional'
};

const localeOptions = Object.keys(locales)
  .map(key => ({
    value: key,
    label: `${key} - ${nameMapper[key] || ''}`
  }))
  .filter(item => nameMapper[item.value]);

const [locale, setLocale] = useState('ja');
const [date, setDate] = useState(null);


<div style={{ display: 'flex', flexFlow: 'column nowrap' }}>
  <select
    style={{ margin: '20px auto' }}
    onChange={e => setLocale(e.target.value)}
    value={locale}
  >
    {localeOptions.map((option, i) => (
      <option value={option.value} key={i}>
        {option.label}
      </option>
    ))}
  </select>
  <Calendar onChange={item => setDate(item)}
 locale={locales[locale]} date={date} />
</div>;
```
