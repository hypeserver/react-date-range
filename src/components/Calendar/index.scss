.rdrCalendarWrapper {
  box-sizing: border-box;
  background: #ffffff;
  display: inline-flex;
  flex-direction: column;
  user-select: none;
}

.rdrDateDisplay{
  display: flex;
  justify-content: space-between;
}

.rdrDateDisplayItem{
  flex: 1 1;
  width: 0;
  text-align: center;
  color: inherit;

  input{
    text-align: inherit;

    &:disabled{
      cursor: default;
    }
  }
}

[dir="ltr"] .rdrDateDisplayItem{
  & + &{
    margin-left: 0.833em;
  }
}

[dir="rtl"] .rdrDateDisplayItem{
  & + &{
    margin-right: 0.833em;
  }
}

.rdrDateDisplayItemActive{}

.rdrMonthAndYearWrapper {
  box-sizing: inherit;
  display: flex;
  justify-content: space-between;
}

.rdrMonthAndYearPickers{
  flex: 1 1 auto;
  display: flex;
  justify-content: center;
  align-items: center;
}

.rdrMonthPicker{}

.rdrYearPicker{}

.rdrNextPrevButton {
  box-sizing: inherit;
  cursor: pointer;
  outline: none;
}

.rdrPprevButton {}

.rdrNextButton {}

.rdrMonths{
  display: flex;
}

.rdrMonthsVertical{
  flex-direction: column;
}

.rdrMonthsHorizontal > div > div > div{
  display: flex;
  flex-direction: row;
}

.rdrMonth{
  width: 27.667em;
}

.rdrWeekDays{
  display: flex;
}

.rdrWeekDay {
  flex-basis: calc(100% / 7);
  box-sizing: inherit;
  text-align: center;
}

.rdrDays{
  display: flex;
  flex-wrap: wrap;
}

.rdrDateDisplayWrapper{}

.rdrMonthName{}

.rdrInfiniteMonths{
  overflow: auto;
}