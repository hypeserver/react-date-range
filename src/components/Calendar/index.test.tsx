import React from 'react';
import { shallow } from 'enzyme';

import Calendar from '.';

describe('Calendar tests', () => {
  test('Should render', () => {
    Date.now = jest.fn(() => 1482363367071);
    const wrapper = shallow(
      <Calendar
        weekStartsOn={0}
        onChange={item => {
          console.log('item selected:', item);
        }}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });
});