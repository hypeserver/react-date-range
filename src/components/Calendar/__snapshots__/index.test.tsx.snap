// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar tests Should render 1`] = `
<div
  className="rdrCalendarWrapper"
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
>
  <div
    className="rdrDateDisplayWrapper"
  />
  <div
    className="rdrMonthAndYearWrapper"
    onMouseUp={[Function]}
  >
    <button
      aria-label="previous month button"
      className="rdrNextPrevButton rdrPprevButton"
      onClick={[Function]}
      type="button"
    >
      <i />
    </button>
    <span
      className="rdrMonthAndYearPickers"
    >
      <span
        className="rdrMonthPicker"
      >
        <select
          aria-label="month picker"
          onChange={[Function]}
          value={9}
        >
          <option
            key="0"
            value={0}
          >
            January
          </option>
          <option
            key="1"
            value={1}
          >
            February
          </option>
          <option
            key="2"
            value={2}
          >
            March
          </option>
          <option
            key="3"
            value={3}
          >
            April
          </option>
          <option
            key="4"
            value={4}
          >
            May
          </option>
          <option
            key="5"
            value={5}
          >
            June
          </option>
          <option
            key="6"
            value={6}
          >
            July
          </option>
          <option
            key="7"
            value={7}
          >
            August
          </option>
          <option
            key="8"
            value={8}
          >
            September
          </option>
          <option
            key="9"
            value={9}
          >
            October
          </option>
          <option
            key="10"
            value={10}
          >
            November
          </option>
          <option
            key="11"
            value={11}
          >
            December
          </option>
        </select>
      </span>
      <span
        className="rdrMonthAndYearDivider"
      />
      <span
        className="rdrYearPicker"
      >
        <select
          aria-label="year picker"
          onChange={[Function]}
          value={2021}
        >
          <option
            key="2041"
            value={2041}
          >
            2041
          </option>
          <option
            key="2040"
            value={2040}
          >
            2040
          </option>
          <option
            key="2039"
            value={2039}
          >
            2039
          </option>
          <option
            key="2038"
            value={2038}
          >
            2038
          </option>
          <option
            key="2037"
            value={2037}
          >
            2037
          </option>
          <option
            key="2036"
            value={2036}
          >
            2036
          </option>
          <option
            key="2035"
            value={2035}
          >
            2035
          </option>
          <option
            key="2034"
            value={2034}
          >
            2034
          </option>
          <option
            key="2033"
            value={2033}
          >
            2033
          </option>
          <option
            key="2032"
            value={2032}
          >
            2032
          </option>
          <option
            key="2031"
            value={2031}
          >
            2031
          </option>
          <option
            key="2030"
            value={2030}
          >
            2030
          </option>
          <option
            key="2029"
            value={2029}
          >
            2029
          </option>
          <option
            key="2028"
            value={2028}
          >
            2028
          </option>
          <option
            key="2027"
            value={2027}
          >
            2027
          </option>
          <option
            key="2026"
            value={2026}
          >
            2026
          </option>
          <option
            key="2025"
            value={2025}
          >
            2025
          </option>
          <option
            key="2024"
            value={2024}
          >
            2024
          </option>
          <option
            key="2023"
            value={2023}
          >
            2023
          </option>
          <option
            key="2022"
            value={2022}
          >
            2022
          </option>
          <option
            key="2021"
            value={2021}
          >
            2021
          </option>
          <option
            key="2020"
            value={2020}
          >
            2020
          </option>
          <option
            key="2019"
            value={2019}
          >
            2019
          </option>
          <option
            key="2018"
            value={2018}
          >
            2018
          </option>
          <option
            key="2017"
            value={2017}
          >
            2017
          </option>
          <option
            key="2016"
            value={2016}
          >
            2016
          </option>
          <option
            key="2015"
            value={2015}
          >
            2015
          </option>
          <option
            key="2014"
            value={2014}
          >
            2014
          </option>
          <option
            key="2013"
            value={2013}
          >
            2013
          </option>
          <option
            key="2012"
            value={2012}
          >
            2012
          </option>
          <option
            key="2011"
            value={2011}
          >
            2011
          </option>
          <option
            key="2010"
            value={2010}
          >
            2010
          </option>
          <option
            key="2009"
            value={2009}
          >
            2009
          </option>
          <option
            key="2008"
            value={2008}
          >
            2008
          </option>
          <option
            key="2007"
            value={2007}
          >
            2007
          </option>
          <option
            key="2006"
            value={2006}
          >
            2006
          </option>
          <option
            key="2005"
            value={2005}
          >
            2005
          </option>
          <option
            key="2004"
            value={2004}
          >
            2004
          </option>
          <option
            key="2003"
            value={2003}
          >
            2003
          </option>
          <option
            key="2002"
            value={2002}
          >
            2002
          </option>
          <option
            key="2001"
            value={2001}
          >
            2001
          </option>
          <option
            key="2000"
            value={2000}
          >
            2000
          </option>
          <option
            key="1999"
            value={1999}
          >
            1999
          </option>
          <option
            key="1998"
            value={1998}
          >
            1998
          </option>
          <option
            key="1997"
            value={1997}
          >
            1997
          </option>
          <option
            key="1996"
            value={1996}
          >
            1996
          </option>
          <option
            key="1995"
            value={1995}
          >
            1995
          </option>
          <option
            key="1994"
            value={1994}
          >
            1994
          </option>
          <option
            key="1993"
            value={1993}
          >
            1993
          </option>
          <option
            key="1992"
            value={1992}
          >
            1992
          </option>
          <option
            key="1991"
            value={1991}
          >
            1991
          </option>
          <option
            key="1990"
            value={1990}
          >
            1990
          </option>
          <option
            key="1989"
            value={1989}
          >
            1989
          </option>
          <option
            key="1988"
            value={1988}
          >
            1988
          </option>
          <option
            key="1987"
            value={1987}
          >
            1987
          </option>
          <option
            key="1986"
            value={1986}
          >
            1986
          </option>
          <option
            key="1985"
            value={1985}
          >
            1985
          </option>
          <option
            key="1984"
            value={1984}
          >
            1984
          </option>
          <option
            key="1983"
            value={1983}
          >
            1983
          </option>
          <option
            key="1982"
            value={1982}
          >
            1982
          </option>
          <option
            key="1981"
            value={1981}
          >
            1981
          </option>
          <option
            key="1980"
            value={1980}
          >
            1980
          </option>
          <option
            key="1979"
            value={1979}
          >
            1979
          </option>
          <option
            key="1978"
            value={1978}
          >
            1978
          </option>
          <option
            key="1977"
            value={1977}
          >
            1977
          </option>
          <option
            key="1976"
            value={1976}
          >
            1976
          </option>
          <option
            key="1975"
            value={1975}
          >
            1975
          </option>
          <option
            key="1974"
            value={1974}
          >
            1974
          </option>
          <option
            key="1973"
            value={1973}
          >
            1973
          </option>
          <option
            key="1972"
            value={1972}
          >
            1972
          </option>
          <option
            key="1971"
            value={1971}
          >
            1971
          </option>
          <option
            key="1970"
            value={1970}
          >
            1970
          </option>
          <option
            key="1969"
            value={1969}
          >
            1969
          </option>
          <option
            key="1968"
            value={1968}
          >
            1968
          </option>
          <option
            key="1967"
            value={1967}
          >
            1967
          </option>
          <option
            key="1966"
            value={1966}
          >
            1966
          </option>
          <option
            key="1965"
            value={1965}
          >
            1965
          </option>
          <option
            key="1964"
            value={1964}
          >
            1964
          </option>
          <option
            key="1963"
            value={1963}
          >
            1963
          </option>
          <option
            key="1962"
            value={1962}
          >
            1962
          </option>
          <option
            key="1961"
            value={1961}
          >
            1961
          </option>
          <option
            key="1960"
            value={1960}
          >
            1960
          </option>
          <option
            key="1959"
            value={1959}
          >
            1959
          </option>
          <option
            key="1958"
            value={1958}
          >
            1958
          </option>
          <option
            key="1957"
            value={1957}
          >
            1957
          </option>
          <option
            key="1956"
            value={1956}
          >
            1956
          </option>
          <option
            key="1955"
            value={1955}
          >
            1955
          </option>
          <option
            key="1954"
            value={1954}
          >
            1954
          </option>
          <option
            key="1953"
            value={1953}
          >
            1953
          </option>
          <option
            key="1952"
            value={1952}
          >
            1952
          </option>
          <option
            key="1951"
            value={1951}
          >
            1951
          </option>
          <option
            key="1950"
            value={1950}
          >
            1950
          </option>
          <option
            key="1949"
            value={1949}
          >
            1949
          </option>
          <option
            key="1948"
            value={1948}
          >
            1948
          </option>
          <option
            key="1947"
            value={1947}
          >
            1947
          </option>
          <option
            key="1946"
            value={1946}
          >
            1946
          </option>
          <option
            key="1945"
            value={1945}
          >
            1945
          </option>
          <option
            key="1944"
            value={1944}
          >
            1944
          </option>
          <option
            key="1943"
            value={1943}
          >
            1943
          </option>
          <option
            key="1942"
            value={1942}
          >
            1942
          </option>
          <option
            key="1941"
            value={1941}
          >
            1941
          </option>
          <option
            key="1940"
            value={1940}
          >
            1940
          </option>
          <option
            key="1939"
            value={1939}
          >
            1939
          </option>
          <option
            key="1938"
            value={1938}
          >
            1938
          </option>
          <option
            key="1937"
            value={1937}
          >
            1937
          </option>
          <option
            key="1936"
            value={1936}
          >
            1936
          </option>
          <option
            key="1935"
            value={1935}
          >
            1935
          </option>
          <option
            key="1934"
            value={1934}
          >
            1934
          </option>
          <option
            key="1933"
            value={1933}
          >
            1933
          </option>
          <option
            key="1932"
            value={1932}
          >
            1932
          </option>
          <option
            key="1931"
            value={1931}
          >
            1931
          </option>
          <option
            key="1930"
            value={1930}
          >
            1930
          </option>
          <option
            key="1929"
            value={1929}
          >
            1929
          </option>
          <option
            key="1928"
            value={1928}
          >
            1928
          </option>
          <option
            key="1927"
            value={1927}
          >
            1927
          </option>
          <option
            key="1926"
            value={1926}
          >
            1926
          </option>
          <option
            key="1925"
            value={1925}
          >
            1925
          </option>
          <option
            key="1924"
            value={1924}
          >
            1924
          </option>
          <option
            key="1923"
            value={1923}
          >
            1923
          </option>
          <option
            key="1922"
            value={1922}
          >
            1922
          </option>
          <option
            key="1921"
            value={1921}
          >
            1921
          </option>
        </select>
      </span>
    </span>
    <button
      aria-label="next month button"
      className="rdrNextPrevButton rdrNextButton"
      onClick={[Function]}
      type="button"
    >
      <i />
    </button>
  </div>
  <div
    className="rdrMonths rdrMonthsVertical"
  >
    <Month
      ariaLabels={
        Object {
          "monthPicker": "month picker",
          "nextButton": "next month button",
          "prevButton": "previous month button",
          "yearPicker": "year picker",
        }
      }
      calendarFocus="forwards"
      classNames={Object {}}
      color="#3d91ff"
      date={2021-10-17T11:38:00.056Z}
      dateDisplayFormat="MMM d, yyyy"
      dateOptions={
        Object {
          "locale": Object {
            "code": "en-US",
            "formatDistance": [Function],
            "formatLong": Object {
              "date": [Function],
              "dateTime": [Function],
              "time": [Function],
            },
            "formatRelative": [Function],
            "localize": Object {
              "day": [Function],
              "dayPeriod": [Function],
              "era": [Function],
              "month": [Function],
              "ordinalNumber": [Function],
              "quarter": [Function],
            },
            "match": Object {
              "day": [Function],
              "dayPeriod": [Function],
              "era": [Function],
              "month": [Function],
              "ordinalNumber": [Function],
              "quarter": [Function],
            },
            "options": Object {
              "firstWeekContainsDate": 1,
              "weekStartsOn": 0,
            },
          },
          "weekStartsOn": 0,
        }
      }
      dayDisplayFormat="d"
      direction="vertical"
      disabledDates={Array []}
      disabledDay={[Function]}
      displayMode="date"
      drag={
        Object {
          "disablePreview": false,
          "range": Object {
            "endDate": null,
            "startDate": null,
          },
          "status": false,
        }
      }
      dragSelectionEnabled={true}
      editableDateInputs={false}
      endDatePlaceholder="Continuous"
      fixedHeight={false}
      focusedRange={
        Array [
          0,
          0,
        ]
      }
      key="0"
      locale={
        Object {
          "code": "en-US",
          "formatDistance": [Function],
          "formatLong": Object {
            "date": [Function],
            "dateTime": [Function],
            "time": [Function],
          },
          "formatRelative": [Function],
          "localize": Object {
            "day": [Function],
            "dayPeriod": [Function],
            "era": [Function],
            "month": [Function],
            "ordinalNumber": [Function],
            "quarter": [Function],
          },
          "match": Object {
            "day": [Function],
            "dayPeriod": [Function],
            "era": [Function],
            "month": [Function],
            "ordinalNumber": [Function],
            "quarter": [Function],
          },
          "options": Object {
            "firstWeekContainsDate": 1,
            "weekStartsOn": 0,
          },
        }
      }
      maxDate={2041-10-17T11:38:00.056Z}
      minDate={1921-10-17T12:38:00.057Z}
      month={2021-09-30T22:00:00.000Z}
      monthDisplayFormat="MMM yyyy"
      months={1}
      onChange={[Function]}
      onDragSelectionEnd={[Function]}
      onDragSelectionMove={[Function]}
      onDragSelectionStart={[Function]}
      onMouseLeave={[Function]}
      onPreviewChange={[Function]}
      preventSnapRefocus={false}
      rangeColors={
        Array [
          "#3d91ff",
          "#3ecf8e",
          "#fed14c",
        ]
      }
      ranges={Array []}
      scroll={
        Object {
          "enabled": false,
        }
      }
      showDateDisplay={true}
      showMonthAndYearPickers={true}
      showMonthArrow={true}
      showMonthName={false}
      showPreview={true}
      showWeekDays={true}
      startDatePlaceholder="Early"
      styles={
        Object {
          "calendarWrapper": "rdrCalendarWrapper",
          "dateDisplay": "rdrDateDisplay",
          "dateDisplayItem": "rdrDateDisplayItem",
          "dateDisplayItemActive": "rdrDateDisplayItemActive",
          "dateDisplayWrapper": "rdrDateDisplayWrapper",
          "dateRangePickerWrapper": "rdrDateRangePickerWrapper",
          "dateRangeWrapper": "rdrDateRangeWrapper",
          "day": "rdrDay",
          "dayActive": "rdrDayActive",
          "dayDisabled": "rdrDayDisabled",
          "dayEndOfMonth": "rdrDayEndOfMonth",
          "dayEndOfWeek": "rdrDayEndOfWeek",
          "dayEndPreview": "rdrDayEndPreview",
          "dayHovered": "rdrDayHovered",
          "dayInPreview": "rdrDayInPreview",
          "dayNumber": "rdrDayNumber",
          "dayPassive": "rdrDayPassive",
          "daySelected": "rdrDaySelected",
          "dayStartOfMonth": "rdrDayStartOfMonth",
          "dayStartOfWeek": "rdrDayStartOfWeek",
          "dayStartPreview": "rdrDayStartPreview",
          "dayToday": "rdrDayToday",
          "dayWeekend": "rdrDayWeekend",
          "days": "rdrDays",
          "definedRangesWrapper": "rdrDefinedRangesWrapper",
          "endEdge": "rdrEndEdge",
          "inRange": "rdrInRange",
          "infiniteMonths": "rdrInfiniteMonths",
          "inputRange": "rdrInputRange",
          "inputRangeInput": "rdrInputRangeInput",
          "inputRangeLabel": "rdrInputRangeLabel",
          "inputRanges": "rdrInputRanges",
          "month": "rdrMonth",
          "monthAndYearDivider": "rdrMonthAndYearDivider",
          "monthAndYearPickers": "rdrMonthAndYearPickers",
          "monthAndYearWrapper": "rdrMonthAndYearWrapper",
          "monthName": "rdrMonthName",
          "monthPicker": "rdrMonthPicker",
          "months": "rdrMonths",
          "monthsHorizontal": "rdrMonthsHorizontal",
          "monthsVertical": "rdrMonthsVertical",
          "nextButton": "rdrNextButton",
          "nextPrevButton": "rdrNextPrevButton",
          "prevButton": "rdrPprevButton",
          "selected": "rdrSelected",
          "startEdge": "rdrStartEdge",
          "staticRange": "rdrStaticRange",
          "staticRangeLabel": "rdrStaticRangeLabel",
          "staticRangeSelected": "rdrStaticRangeSelected",
          "staticRanges": "rdrStaticRanges",
          "weekDay": "rdrWeekDay",
          "weekDays": "rdrWeekDays",
          "yearPicker": "rdrYearPicker",
        }
      }
      weekStartsOn={0}
      weekdayDisplayFormat="E"
    />
  </div>
</div>
`;
